### PR TITLE
 Admin area codemirror JSON editor does not display arrays

### PR DESCRIPTION
**Pull request by tazo90**
It allows to display arrays in codemirror editor.

### CODE REVIEW

**Comment by lsbardel** [comment]
@tazo90 can you update this pull request so that @reupen can review?

**Comment by tazo90** [comment]
@lsbardel I've updated PR.

**Comment by reupen** [comment]
Hi @tazo90,

This problem isn't too important at the moment, but comments are:

* A JSON field can be a string, number, boolean, array or 'object'. There is an existing check [here](https://github.com/quantmind/lux/pull/171/files#diff-c53c6a7f4f24f5092ab9d33efb772e59L95) similar to the new one. Instead of adding another check for arrays we should modify the existing one to stringify any JSON field value, not just objects. We probably need to modify [this line in forms.js](https://github.com/quantmind/lux/blob/d076f57158d6e43e2e57eac5bf0b4cd6ccbadbe1/luxjs/forms/form.js#L162) to store all the field info so we know whether it is a JSON field or not:
* The other problem is that the back end is appending '[]' to the field name if the value of the JSON field is an array. I don't think it should be doing this as otherwise the name of the field changes depending on the value!

Thanks

**Comment by tazo90** [comment]
@reupen thanks for your suggestions, it makes sense. I will do changes.

**Comment by lsbardel** [comment]
@reupen right, we don't want that. I was not aware of that bug.

**Comment by lsbardel** [comment]
Can this be tested?

**Comment by tazo90** [comment]
@lsbardel I will add tests for this obviously.

**Comment by tazo90** [comment]
Fixed.

This removes the braces from json arrays fields because of following issue [https://github.com/quantmind/lux/issues/197](https://github.com/quantmind/lux/issues/197).

- [x] fixed codemirror issue for json arrays 
- [x] tests

**Preview**
json-arrays

File: luxjs/utils.js [hunk 1]
```diff
@@ -245,6 +245,20 @@
     },
     //
     //  Capitalize the first letter of string
-    capitalize = function(str) {
+    capitalize = function (str) {
         return str.charAt(0).toUpperCase() + str.slice(1);
+    },
+
+    /**
+     * Obtain an json object from string (if available) otherwise null
```
Comment (lsbardel) [inline luxjs/utils.js]: Obtain a JSON object from a string (if available) otherwise null

File: luxjs/tests/lux.js [hunk 2]
```diff
@@ -86,4 +86,12 @@
             expect(body.length).toBe(1);
             expect(body[0].tagName).toBe('BODY');
         });
+
+        it("Check getJsonOrNone", function() {
+            var res;
+            res = lux.getJsonOrNone("{\"id\": 1}");
+            expect(res.id).toBeDefined();
```
Comment (lsbardel) [inline luxjs/tests/lux.js]: ``.toBe(1)``

**Comment by tazo90** [comment]
Fixed.

File: luxjs/services/luxweb.js [hunk 3]
```diff
@@ -92,19 +92,28 @@
                             // TODO: do we need a callback for JSON fields?
                             // or shall we leave it here?
 
-                            if (formScope[formScope.formModelName + 'Type'][key] === 'textarea' && isObject(value)) {
+                            var modelType = formScope[formScope.formModelName + 'Type'];
+                            var jsonArrayKey = key.split('[]')[0];
+
+                            // Stringify json only if has json mode enabled
+                            if (modelType[jsonArrayKey] === 'json' && (isObject(value) || isArray(value))) {
```
Comment (reupen) [inline luxjs/services/luxweb.js]: Strings must be JSON-encoded too (for a JSON field).

Comment (tazo90) [inline luxjs/services/luxweb.js]: I added a condition for a Strings too.

**Pull request merged**